pkg/controller/controller_utils.go

```diff
@@ -629,6 +686,135 @@ func maxContainerRestarts(pod *api.Pod) int {
+	controllerKind unversioned.GroupVersionKind,
+) *podControllerRefManager {
+	return &podControllerRefManager{podControl, controllerObject, controllerSelector, controllerKind}
+}
+
+// ClassifyActivePods first filters out inactive pods, then it classify the
+// remaining pods into three categories: 1. matchesAndControlled are the pods
+// whose labels match the selector of the RC, and have a controllerRef pointing
+// to the controller 2. matchesNeedsController are the pods whose labels match
+// the RC, but don't have a controllerRef. (Pods with matching labels but with a
+// controllerRef pointing to other object are ignored) 3. controlledDoesNotMatch
+// are the pods that have a controllerRef pointing to the controller, but their
+// labels no longer match the selector.
+func (m *podControllerRefManager) ClassifyActivePods(pods []api.Pod) (
+	matchesAndControlled []*api.Pod,
+	matchesNeedsController []*api.Pod,
+	controlledDoesNotMatch []*api.Pod) {
+	for i := range pods {
+		pod := pods[i]
+		if !IsPodActive(pod) {
+			glog.V(4).Infof("Ignoring inactive pod %v/%v in state %v, deletion time %v",
+				pod.Namespace, pod.Name, pod.Status.Phase, pod.DeletionTimestamp)
+			continue
+		}
+		controllerRef := getControllerOf(pod.ObjectMeta)
+		if controllerRef != nil {
+			if controllerRef.UID == m.controllerObject.UID {
+				// already controlled
+				if m.controllerSelector.Matches(labels.Set(pod.Labels)) {
+					matchesAndControlled = append(matchesAndControlled, &pod)
+				} else {
+					controlledDoesNotMatch = append(controlledDoesNotMatch, &pod)
+				}
+			} else {
+				// ignoring the pod controlled by other controller
+				glog.V(4).Infof("Ignoring pod %v/%v, it's owned by [%s/%s, name: %s, uid: %s]",
+					pod.Namespace, pod.Name, controllerRef.APIVersion, controllerRef.Kind, controllerRef.Name, controllerRef.UID)
+				continue
+			}
+		} else {
+			if !m.controllerSelector.Matches(labels.Set(pod.Labels)) {
+				continue
+			}
+			matchesNeedsController = append(matchesNeedsController, &pod)
+		}
+	}
+	return matchesAndControlled, matchesNeedsController, controlledDoesNotMatch
+
+}
+
+// getControllerOf returns the controllerRef if controllee has a controller,
+// otherwise returns nil.
+func getControllerOf(controllee api.ObjectMeta) *api.OwnerReference {
+	for _, owner := range controllee.OwnerReferences {
+		// controlled by other controller
+		if owner.Controller != nil && *owner.Controller == true {
+			return &owner
+		}
+	}
+	return nil
+}
+
+// ControlPods sends a patch to take control of the pods. It returns the list of
+// pods that are successfully controlled, and the list of errors that occur in
+// failed patching attempts.
+func (m *podControllerRefManager) ControlPods(pods []*api.Pod) (controlledPods []*api.Pod, errlist []error) {
+	// we should not adopt any pods if the controller is about to be deleted
+	if m.controllerObject.DeletionTimestamp != nil {
+		return nil, nil
+	}
+	for _, pod := range pods {
```

Comment:
I removed the loop. I thought all pod controllers might all need the loop, we can add the loop utility function if that's the case.